core/l10n_init.js

```diff
@@ -29,20 +26,37 @@
             'tr', 'xh', 'zh-CN', 'zh-TW', 'zu', 'dbg'
         ];
     }
-    injectLocaleScript();
 
-    function injectLocaleScript(_testLocale) {
+    function injectLocaleScript(promise, locale) {
         // Inject script to load locale file that populates w.navigator.1l0n.
+        // If locale script 404s, fall back to en-US.
+        promise = promise || defer.Deferred();
+        locale = locale || getLocale();
+
         var script = document.createElement('script');
-        script.src = getLocaleSrc(_testLocale || getLocale());
+        script.src = getLocaleSrc(locale);
+        script.onload = function() {
+            promise.resolve(locale);
+        };
         script.onerror = function() {
-            // If locale file not found, then fallback English (bug 1044195).
-            window.navigator.l10n = {
-                language: 'en-US'
-            };
+            // If locale file not found, then fall back English (bug 1044195).
+            if (locale == 'en-US') {
+                // If we are falling back and it still errors, just resolve.
+                window.navigator.l10n = {
+                    language: 'en-US'
+                };
+                promise.resolve('en-US');
+            } else {
+                // Fall back to en-US by trying to load the en-US locale.
+                injectLocaleScript(promise, 'en-US');
+            }
         };
         document.body.appendChild(script);
-        return script;
+
+        return {
+            promise: promise,
+            script: script
```

Comment:
Rather than returning `script` here should it set the script as a second argument to `resolve` and `reject`?